packages/cli-main/src/cli/services/upgrade.ts

```diff
@@ -0,0 +1,174 @@
+  } else if (usingPackageManager()) {
+    throw new error.Abort(
+      output.content`Couldn't find the configuration file for ${output.token.path(
+        directory,
+      )}, are you in a Shopify project directory?`,
+    )
+  } else {
+    newestVersion = await upgradeGlobalShopify(currentVersion)
+  }
+
+  if (newestVersion) {
+    output.success(`Upgraded Shopify CLI to version ${newestVersion}`)
+  }
+}
+
+async function getProjectDir(directory: string) {
+  return path.findUp(
+    async (dir: string) => {
+      const configFilesExist = await Promise.all(
+        ['shopify.app.toml', 'hydrogen.config.js', 'hydrogen.config.ts'].map(async (configFile) => {
+          return file.exists(path.join(dir, configFile))
+        }),
+      )
+      if (configFilesExist.some((bool) => bool)) return dir
+    },
+    {
+      cwd: directory,
+      type: 'directory',
+    },
+  )
+}
+
+async function upgradeLocalShopify(projectDir: string, currentVersion: string): Promise<string | void> {
+  const packageJson = (await findUpAndReadPackageJson(projectDir)).content as PackageJsonContents
+  const packageJsonDependencies = packageJson.dependencies || {}
+  const packageJsonDevDependencies = packageJson.devDependencies || {}
+
+  let resolvedVersion: string = {...packageJsonDependencies, ...packageJsonDevDependencies}[await cliDependency()]!
+  if (resolvedVersion.slice(0, 1).match(/[\^~]/)) resolvedVersion = currentVersion
+  const newestVersion = await checkForNewVersion(await cliDependency(), resolvedVersion)
+
+  if (!newestVersion) return wontInstall(resolvedVersion)
+
+  outputUpgradeMessage(resolvedVersion, newestVersion)
+
+  await installJsonDependencies('prod', packageJsonDependencies, projectDir)
+  await installJsonDependencies('dev', packageJsonDevDependencies, projectDir)
+  return newestVersion
+}
+
+async function upgradeGlobalShopify(currentVersion: string): Promise<string | void> {
+  const newestVersion = await checkForNewVersion(await cliDependency(), currentVersion)
+
+  if (!newestVersion) return wontInstall(currentVersion)
+
+  outputUpgradeMessage(currentVersion, newestVersion)
+
+  const {platform} = os.platformAndArch()
+  if (platform.match(/darwin/)) {
+    try {
+      const brewList = await system.captureOutput('brew', ['list', '-1'])
+      if (brewList.match(/^shopify-cli@3$/m)) {
+        output.info(
+          output.content`Homebrew installation detected. Attempting to upgrade via ${output.token.genericShellCommand(
+            'brew upgrade',
+          )}...`,
+        )
+        await system.exec('brew', ['upgrade', 'shopify-cli@3'], {stdio: 'inherit'})
+        return newestVersion
+      }
+      // eslint-disable-next-line no-catch-all/no-catch-all
+    } catch (err) {
+      output.warn('Homebrew upgrade failed. Falling back to standard npm install.')
+    }
+  }
+  const command = 'npm'
+  const args = [
+    'install',
+    '-g',
+    `${await cliDependency()}@latest`,
```

Comment:
I see more than one `await cliDependency()` scattered in this function. What about calling it once and passing it down?